### ISSUE TITLE
Can't mock a method that returns a reference to a non-Send type

### ISSUE THREAD

**Issue by asomers** [asomers/mockall#9]
Mockall can mock methods that return non-`Send` types, but it can't currently mock methods that return references to non-`Send` types.  For example, this won't work:
```rust
use mockall::*;

mock! {
    Foo {
        fn baz(&self) -> &std::rc::Rc<u32>;
    }
}
```
```
error[E0277]: `std::rc::Rc<u32>` cannot be shared between threads safely
  --> mockall/tests/mock_return_reference.rs:6:1
   |
6  | / mock! {
7  | |     Foo {
8  | |         fn baz(&self) -> &std::rc::Rc<u32>;
9  | |     }
10 | | }
   | |_^ `std::rc::Rc<u32>` cannot be shared between threads safely
   |
   = help: the trait `std::marker::Sync` is not implemented for `std::rc::Rc<u32>`
   = help: see issue #48214
   = help: add #![feature(trivial_bounds)] to the crate attributes to enable
```

**Comment by asomers** [comment]
Note that methods returning mutable references are affected too.  For example:
```rust
use mockall::*;

trait Foo {
    fn foo<I: 'static, O: 'static>(&self, i: I) -> &mut O;
}

mock!{
    MyStruct {}
    trait Foo {
        fn foo<I: 'static, O: 'static>(&self, i: I) -> &mut O;
    }
}
```

**Comment by necnec** [comment]
Hi @asomers !
Very inspired by your lib. I think I have a similar issue. Is there any workaround to mock functions which return reference like `&[String]`?
```rust
pub trait ModelData {
    fn foo(&self, field: &str) -> Option<String>;
}


use cfg_if::cfg_if;

cfg_if! {
    if #[cfg(test)] {
        use mockall::*;
        mock! {
            Data {}
            trait ModelData {
                fn foo(&mut self) -> &[String];
            }
        }
    }
}
```
Which produces following error message:
```rust
error[E0277]: the size for values of type `[std::string::String]` cannot be known at compilation time
   --> src/tool/data.rs:104:65
    |
104 |                 fn foo(&mut self) -> &[String];
    |                                       ^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `[std::string::String]`
    = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
    = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the size for values of type `[std::string::String]` cannot be known at compilation time
   --> src/tool/data.rs:98:9
    |
98  | /         mock! {
99  | |             Data {}
100 | |             trait ModelData {
101 | |                 fn get_field(&self, field: &str) -> Option<String>;
...   |
144 | |             }
145 | |         }
    | |_________^ doesn't have a size known at compile-time
    |
    = help: within `tool::data::__mock_Data_ModelData::foo::Rfunc`, the trait `std::marker::Sized` is not implemented for `[std::string::String]`
    = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
    = note: required because it appears within the type `tool::data::__mock_Data_ModelData::foo::Rfunc`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

```

**Comment by asomers** [comment]
I'm confused.  In your example, the signature for `foo` in the `mock!` does not match the signature in the definition of the trait.  Why not?

Assuming that's just a copy/paste error, there is indeed an easy workaround for your problem.  It has nothing to do with `Send`.  The problem is that when you set the expectation for a mock method that returns a reference, you must supply an owned value of the type referred.  That method will look like this:
```rust
fn return_const(&mut self, var: [String])
```
That's why you're seeing the error, because `[String]` is unsized, it can't be used as a function argument.  Here are two possible workarounds:

1) Change the return type to be a  `&'static [String]`.  Then you won't be expected to provide any owned values.

2) Change the return type to be a sized type, like `&Vec<String>`.  That way the owned version will be `Sized`, but the caller will still be able to manipulate it just as if it were a slice.

Hopefully one of these options will work for you.  If not, please open a new issue.

**Comment by necnec** [comment]
Thanks! using `&Vec<String>` solves the issue.

> In your example, the signature for foo in the mock! does not match the signature in the definition of the trait. Why not?

Sorry for that, just wanted to throw out all extra stuff, now example is updated

>  It has nothing to do with Send

Sorry for that, I thought to create separate issue, but error seemed too similar

**Comment by asomers** [comment]
I inadvertently fixed this issue sometime  between 0.4.0 and 0.5.0.
